File: 2026/06-June/27-Saturday/solution.py
from collections import Counter

class Solution:
    def maximumLength(self, nums: list[int]) -> int:
        count = Counter(nums)
        
        # Handle edge case for 1: max odd number of 1s
        ans = count[1] - (count[1] % 2 == 0) if count[1] > 0 else 1
        
        for num in count:
            if num == 1:
                continue
                
            length = 0
            x = num
            
            # As long as we have at least 2 of the current number, we can extend both sides
            while count[x] >= 2:
                length += 2
                x *= x
            
            # Check if the next squared element can serve as the single peak element
            if count[x] >= 1:
                length += 1
            else:
                # If it doesn't exist, the last element we counted twice must act as the peak
                length -= 1
                
            ans = max(ans, length)
            
        return ans

File: 2026/06-June/27-Saturday/solution.java
import java.util.*;

class Solution {
    public int maximumLength(int[] nums) {
        Map<Integer, Integer> count = new HashMap<>();
        int maxNum = 0;
        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
            maxNum = Math.max(maxNum, num);
        }
        
        // Handle edge case for 1
        int ans = 1;
        if (count.containsKey(1)) {
            int ones = count.get(1);
            ans = (ones % 2 == 0) ? ones - 1 : ones;
        }
        
        for (int num : count.keySet()) {
            if (num == 1) continue;
            
            int length = 0;
            long x = num; // Use long to prevent integer overflow during squaring
            
            while (x <= maxNum && count.getOrDefault((int)x, 0) >= 2) {
                length += 2;
                x = x * x;
            }
            
            if (x <= maxNum && count.getOrDefault((int)x, 0) >= 1) {
                length += 1;
            } else {
                length -= 1;
            }
            
            ans = Math.max(ans, length);
        }
        
        return ans;
    }
}
